### ISSUE TITLE
Allow set font application-wide

### ISSUE THREAD

**Issue by RussKie** [dotnet/winforms#3001]
**Is your feature request related to a problem? Please describe.**

The default font has been updated in .NET Core 3.0 (#656) and [documented](https://docs.microsoft.com/en-us/dotnet/core/compatibility/fx-core#default-control-font-changed-to-segoe-ui-9-pt). However for some users there is still an element of surprise when they migrate their apps to .NET Core.
We've received several questions regarding different sizes of forms (e.g. https://github.com/dotnet/winforms/issues/1122, https://github.com/dotnet/winforms/issues/1827, etc.).

Whilst the new default font is here to stay, some users may wish to retain the original font (e.g. due to a design of their app). However for an application with more than a handful of forms, setting the original font may be tedious and cumbersome exercise.

**Describe the solution you'd like**

Add the ability to set an application-wide font, similar to `SetHighDpiMode()` or `SetCompatibleTextRenderingDefault()` methods.

* This method must be "run only once" kind, i.e. a user may not be allowed to invoke it once an app has started.
* Any form that doesn't explicitly specify its own font, must inherit the application default font.

### API Proposal

```C#
namespace System.Windows.Forms
{
    public partial class Application
    {
        public void SetDefaultFont(Font font);
    }
}
```

### API Usage

```C#
class Program
{
    [STAThread]
    static void Main()
    {
        Application.SetHighDpiMode(HighDpiMode.SystemAware);
        Application.EnableVisualStyles();
        Application.SetCompatibleTextRenderingDefault(false);

        Application.SetDefaultFont(new Font(new FontFamily("Microsoft Sans Serif"), 8f));

        Application.Run(new Form1());
    }
}
```

**Will this feature affect UI controls?**

No

**Comment by merriemcgaw** [comment]
We should totally do this for .NET 5, I think this is a great idea.

**Comment by terrajobst** [comment]
> * This method must be "run only once" kind, i.e. a user may not be allowed to invoke it once an app has started.

Are you going to throw an exception when the `SetDefaultFont` is called after `Run`?

**Comment by RussKie** [comment]
> > * This method must be "run only once" kind, i.e. a user may not be allowed to invoke it once an app has started.
> 
> Are you going to throw an exception when the `SetDefaultFont` is called after `Run`? If so, I'd throw `InvalidOperationException`.

I propose we follow the same pattern set by `SetCompatibleTextRenderingDefault()` method and throw if `SetDefaultFont()` is called after the application has started:
https://github.com/dotnet/winforms/blob/18f1c8101085fd60a3eda7c95c796edc86c65820/src/System.Windows.Forms/src/System/Windows/Forms/Application.cs#L1208-L1215

The rationale behind it, if a form has been created, it may have been laid out and rendered using one font, the font may have been cached etc., and setting a new form could be detrimental from both performance and UX.

**Comment by bartonjs** [comment]
Is there already a way to get the default font?  (Get/set symmetry)

**Comment by weltkante** [comment]
probably `Control.DefaultFont` ?

**Comment by terrajobst** [comment]
Might be worthwhile to add a `GetDefaultFont()` so that folks can just tweak parts, like:

```C#
[STAThread]
static void Main()
{
    Application.SetHighDpiMode(HighDpiMode.SystemAware);
    Application.EnableVisualStyles();
    Application.SetCompatibleTextRenderingDefault(false);

    // Set default to 10pt
    var font = Application.GetDefaultFont();
    font.SetSize(10f);
    Application.SetDefaultFont(font);

    Application.Run(new Form1());
}
```

**Comment by weltkante** [comment]
I don't think fonts are editable, you have to construct a new one.

**Comment by terrajobst** [comment]
> I don't think fonts are mutable, you have to construct a new one. (Not arguing against Application.GetDefaultFont for discoverability.)

Fine, be like that ;-) Fixed sample.

**Comment by RussKie** [comment]
The rationale behind this FR is to allow users to set the original font, that may have been used to design the app (i.e. _Font-scaling pixel-perfect_ design scenario), to help with a migration from .NET Framework to .NET Core/.NET. 
To this end, users are expected to set both `FontFamily` and `FontSize` explicitly, and not leverage the current default font.

I don't feel, that leveraging the default font to change its properties fits the above scenario. And, to be honest, I'm struggling to think of a use-case that would require this. 
Do you have a use-case that would necessitate this?

**Comment by RussKie** [comment]
Bump 😉

**Comment by terrajobst** [comment]
No objections, you're the domain expert. We just generally strive to avoid asymmetries like set-only APIs.

**Comment by RussKie** [comment]
> We just generally strive to avoid asymmetries like set-only APIs.

So do we, but in this case we want to have set-once kind of API, akin to `SetHighDpiMode()` or `EnableVisualStyles()`.
Thank you

**Comment by terrajobst** [comment]
Reviewed and approved offline:

```C#
namespace System.Windows.Forms
{
    public partial class Application
    {
        public void SetDefaultFont(Font font);
    }
}
```

**Comment by RussKie** [comment]
We had further team discussions about the API and how it will work with the designer, and we identified a number of gaps, which require further analysis.
I'm going to close it until we fully establish this is the correct approach.

**Comment by RussKie** [comment]
We can now implement this proposal. https://github.com/dotnet/winforms/issues/4909 will facilitate a communication of the the font information to the designer process.

**Comment by kirsan31** [comment]
@RussKie #4009 is grate! But,  I want to point out some one thing that related to this issue and #4009 too. It's system font scaling:
![image](https://user-images.githubusercontent.com/17767561/118230601-9a321680-b496-11eb-8cb3-9c878537f949.png)
Currently explicitly set font of the control (most time in designer), will stop it's scaling with this technique. What will be if we set _app wide_ font?
My thoughts on this:
Current behavior is not good at all. The `System.Windows.Forms.AutoScaleMode.Font` is default to the form and disabling this scale if you set the font is not obvious.
I want an ability to set the font and still be the part of system font scaling (at least if we use default font but different size / style). #3263 related. Good question what to do if we use not default (system) font at all? But in 90% of time we need a proportional increase of all controls, regardless of the font. For other 10% we have `AutoScaleMode.Dpi` or `AutoScaleMode.None`.

**Comment by RussKie** [comment]
> It's system font scaling:
> ![image](https://user-images.githubusercontent.com/17767561/118230601-9a321680-b496-11eb-8cb3-9c878537f949.png)
> Currently explicitly set font of the control (most time in designer), will stop it's scaling with this technique. What will be if we set _app wide_ font?

Good call!

I've created a simple app targeting net472 and net6.0
```cs
        public Form1()
        {
            InitializeComponent();
            
            Text = Control.DefaultFont.ToString();
            
            DoubleClick += (s, e) => Text = "! " + Control.DefaultFont.ToString();
        }
```

Changing the text size I didn't observe the any changes to the font metrics in net472:

* 100%
![image](https://user-images.githubusercontent.com/4403806/118384554-3fd9b700-b64a-11eb-9b00-29a44c68249d.png)

* 120%
![image](https://user-images.githubusercontent.com/4403806/118384548-25074280-b64a-11eb-942e-491c62040dc9.png)
![image](https://user-images.githubusercontent.com/4403806/118384529-05701a00-b64a-11eb-9d35-f28cb74c5e4b.png)

...but net6.0 appears to be updating the default font metrics:
* 100%
![image](https://user-images.githubusercontent.com/4403806/118384567-60a20c80-b64a-11eb-97fa-6d6c30144f6d.png)
* 120%
![image](https://user-images.githubusercontent.com/4403806/118384592-8b8c6080-b64a-11eb-8981-219f999f7e01.png)

**Comment by mdtauk** [comment]
With Windows 10 21H2 changing the System font to **Segoe UI Variable** with its **Small**, **Text**, and **Display** OpticalSize variants - will WinForms be ready to handle that, and will there be suitable fallbacks for Windows 8, and older Windows 10 versions?

https://blogs.windows.com/windows-insider/2021/05/06/announcing-windows-10-insider-preview-build-21376/

**Comment by kirsan31** [comment]
@RussKie 

> Changing the text size I didn't observe the any changes to the font metrics in net472
> ...
> but net6.0 appears to be updating the default font metrics

Yes, becouse .net framework doesn't respect system font, and .net core does (#656). But font scaling in .net will work only if you not set font explicitly (then it will be taken from `Control.DefaultFont`). Thats all my previous [post](https://github.com/dotnet/winforms/issues/3001#issuecomment-841059077) about.
Intresting that your .net6 120% screen is the same size as 100% - font scailing not working..

**Comment by RussKie** [comment]
> And I see that size of the form is the same on all of your screenshots, it means that font scaling **not working**...

They are in fact different in net6.0 scenarios :)

**Comment by RussKie** [comment]
> With Windows 10 21H2 changing the System font to **Segoe UI Variable** with its **Small**, **Text**, and **Display** OpticalSize variants - will WinForms be ready to handle that, and will there be suitable fallbacks for Windows 8, and older Windows 10 versions?

Thanks for the link. We get the default font from .NET runtime, and take it for granted:
https://github.com/dotnet/winforms/blob/c93d8c85ccd53597b6c9dfe83637ada82fabe353/src/System.Windows.Forms/src/System/Windows/Forms/Control.cs#L1807-L1819

If you're running that build already you could probably run a test yourself and see:
https://github.com/dotnet/runtime/blob/64303750a9198a49f596bcc3aa13de804e421579/src/libraries/System.Drawing.Common/src/System/Drawing/SystemFonts.Windows.cs#L85-L99

I think it may just work out of the box... @safern any thoughts?

**Comment by RussKie** [comment]
> P.s. I want to remind some old annoying [DataGridView Designer issue](https://developercommunity2.visualstudio.com/t/WinForms-DataGridView-styles-designer/1143911#T-N1323376), that become more annoying due to working font scaling in .net. And I think it definitely needs to be addressed 😋

I've tested this in, and the font serialization has been fixed in .NET 5.0 (in https://github.com/dotnet/winforms/pull/3875). 
However looks like the CellStyle editor is settings the default font to "Microsoft Sans Serif, 8.25f", this is on the designer side (I think).
![dgv-cellstyle-font](https://user-images.githubusercontent.com/4403806/118398742-578e5b00-b69d-11eb-9181-35f1045fb1f3.gif)

**Comment by kirsan31** [comment]
> > P.s. I want to remind some old annoying [DataGridView Designer issue](https://developercommunity2.visualstudio.com/t/WinForms-DataGridView-styles-designer/1143911#T-N1323376), that become more annoying due to working font scaling in .net. And I think it definitely needs to be addressed 😋
> 
> I've tested this in, and the font serialization has been fixed in .NET 5.0 (in #3875).
> However looks like the CellStyle editor is settings the default font to "Microsoft Sans Serif, 8.25f", this is on the designer side (I think).

Yea it's all about designer, I [answered](https://developercommunity2.visualstudio.com/t/WinForms-DataGridView-styles-designer/1143911#T-N1424480) to your post on developercommunity, thank you.

**Comment by Zheng-Li01** [comment]
Verified the issue with latest 6.0.0-preview.5.21275.7, the main functions have been fully merged/implemented, found a new issue too and filed it as #[4978](https://github.com/dotnet/winforms/issues/4978)

**Comment by imba-tjd** [comment]
Question: Can WPF set default font? rather than set FontFamily in every xaml.

**Comment by weltkante** [comment]
@imba-tjd you can try setting the default value of the FontFamily and related properties in the metadata before creating any window or control. In general this is the wrong repo to ask in though, you'll find more people knowing about WPF in the WPF repo.
